### ISSUE TITLE
Both memberIds.gender and memberIds found in fields option, using both of them may trigger unexpected behavior.

### ISSUE THREAD

**Issue by crapthings** [cult-of-coders/grapher#78]
```
Exception from sub exposure_projects id TJPSKaKHvqj2j8Jn7 MinimongoError: both memberIds.gender and memberIds found in fields option, using both of them may trigger unexpected behavior. Did you mean to use only one of them?
```

owner = Meteor.users.findOne
members = Meteor.users.find

> query
```
{
    "name": 1,
    "code": 1,
    "type": 1,
    "establishedAt": 1,
    "owner": {
        "name": 1,
        "organization": {
            "name": 1
        }
    },
    "memberIds": {
        "name": 1,
        "gender": 1
    },
    "members": {
        "name": 1,
        "gender": 1
    }
}
```

i think if field is from the same collection, it should merge.

**Comment by crapthings** [comment]
```
Projects.addLinks({
    owner: {
        type: 'one',
        collection: Meteor.users,
        field: 'ownerId',
    },

    members: {
        type: 'many',
        collection: Meteor.users,
        field: 'memberIds',
    },
})
```

when query   will cause MinimongoError: both memberIds.gender and memberIds found in fields option, using both of them may trigger unexpected behavior. Did you mean to use only one of them?

**Comment by theodorDiaconu** [comment]
@crapthings this is related to mongo, basically your query transforms the fields in something like:
```
{
       memberIds: 1,
       'memberIds.gender': 1
}
```

Which is ofcourse redundant. I assume your members need metadata, since you want to store aditional information such as "gender". Make sure you read about many-meta links.

This $metadata, will be stored in the children when you do `{ projects: { members : {} } }`

**Comment by theodorDiaconu** [comment]
Please re-open if it doesn't solve your issue.

**Comment by crapthings** [comment]
oh  i think i've found the problem, this only reproduce #77 

```
https://github.com/crapthings/grapher-reproduce-101
```

**Comment by crapthings** [comment]
something must be wrong

```
{
    "name": 1,
    "code": 1,
    "type": 1,
    "establishedAt": 1,
    "owner": {
        "name": 1,
        "organization": {
            "name": 1
        }
    },
    "memberIds": 1,
    "members": {
        "name": 1,
        "gender": 1
    }
}
```

```
{
        "code": "编号 1407",
        "establishedAt": "2016-03-25T06:19:22.215Z",
        "memberIds": [
            "7M2MthqpiMry95Xs7",
            "cDgGQvkJpostf3iTp",
            "CP4oZoh5JmHnKWs4Z",
            "7SMEkxMfsNJsw5Dxz",
            "LgKBonRAKme47g54t",
            "k8sDjGHKHhAQe8T9s",
            "3ZRP9KzFsS9wHphqN",
            "XWLNBorY2XLMxdxpT"
        ],
        "name": "项目 9261",
        "type": "方式1",
        "_id": "Hfth5WbtGPfXgaCrC",
        "owner": {
            "name": "赵畅",
            "_id": "CP4oZoh5JmHnKWs4Z"
        },
        "members": [
            {
                "gender": "女",
                "name": "赵畅",
                "_id": "CP4oZoh5JmHnKWs4Z"
            },
            {
                "gender": "女",
                "name": "刘菲",
                "_id": "3ZRP9KzFsS9wHphqN"
            },
            {
                "gender": "男",
                "name": "高菲",
                "_id": "7M2MthqpiMry95Xs7"
            },
            {
                "gender": "女",
                "name": "向宏",
                "_id": "7SMEkxMfsNJsw5Dxz"
            },
            {
                "gender": "男",
                "name": "刘宝强",
                "_id": "LgKBonRAKme47g54t"
            },
            {
                "gender": "男",
                "name": "向博",
                "_id": "XWLNBorY2XLMxdxpT"
            },
            {
                "gender": "男",
                "name": "向朋友",
                "_id": "cDgGQvkJpostf3iTp"
            },
            {
                "gender": "女",
                "name": "周大海",
                "_id": "k8sDjGHKHhAQe8T9s"
            }
        ]
    },
```

owner's organization field failed to fetch

**Comment by crapthings** [comment]
if i remove that gender in query, it works.

```
{
    "name": 1,
    "code": 1,
    "type": 1,
    "establishedAt": 1,
    "owner": {
        "name": 1,
        "organization": {
            "name": 1
        }
    },
    "memberIds": 1,
    "members": {
        "name": 1
    }
}
```

```
{
        "code": "编号 1407",
        "establishedAt": "2016-03-25T06:19:22.215Z",
        "memberIds": [
            "7M2MthqpiMry95Xs7",
            "cDgGQvkJpostf3iTp",
            "CP4oZoh5JmHnKWs4Z",
            "7SMEkxMfsNJsw5Dxz",
            "LgKBonRAKme47g54t",
            "k8sDjGHKHhAQe8T9s",
            "3ZRP9KzFsS9wHphqN",
            "XWLNBorY2XLMxdxpT"
        ],
        "name": "项目 9261",
        "type": "方式1",
        "_id": "Hfth5WbtGPfXgaCrC",
        "owner": {
            "name": "赵畅",
            "_id": "CP4oZoh5JmHnKWs4Z",
            "organization": {
                "name": "大庆苹果集团",
                "_id": "vxu7QaYhLLQcwFDCZ"
            }
        },
        "members": [
            {
                "name": "赵畅",
                "_id": "CP4oZoh5JmHnKWs4Z"
            },
            {
                "name": "刘菲",
                "_id": "3ZRP9KzFsS9wHphqN"
            },
            {
                "name": "高菲",
                "_id": "7M2MthqpiMry95Xs7"
            },
            {
                "name": "向宏",
                "_id": "7SMEkxMfsNJsw5Dxz"
            },
            {
                "name": "刘宝强",
                "_id": "LgKBonRAKme47g54t"
            },
            {
                "name": "向博",
                "_id": "XWLNBorY2XLMxdxpT"
            },
            {
                "name": "向朋友",
                "_id": "cDgGQvkJpostf3iTp"
            },
            {
                "name": "周大海",
                "_id": "k8sDjGHKHhAQe8T9s"
            }
        ]
    },
```

**Comment by crapthings** [comment]
okay, after test, i can confirm

if i query some fields that linked to same collection,
i have to write all fields that i don't need it in owner but in members.

when the query object doesnt equal, the nest relation will lost.

sorry for my english


```
{
    "name": 1,
    "code": 1,
    "type": 1,
    "establishedAt": 1,
    "owner": {
        "name": 1,
        "gender": 1,
        "organization": {
            "name": 1
        }
    },
    "memberIds": 1,
    "members": {
        "name": 1,
        "gender": 1
    }
}
```

```
{
        "code": "编号 1407",
        "establishedAt": "2016-03-25T06:19:22.215Z",
        "memberIds": [
            "7M2MthqpiMry95Xs7",
            "cDgGQvkJpostf3iTp",
            "CP4oZoh5JmHnKWs4Z",
            "7SMEkxMfsNJsw5Dxz",
            "LgKBonRAKme47g54t",
            "k8sDjGHKHhAQe8T9s",
            "3ZRP9KzFsS9wHphqN",
            "XWLNBorY2XLMxdxpT"
        ],
        "name": "项目 9261",
        "type": "方式1",
        "_id": "Hfth5WbtGPfXgaCrC",
        "owner": {
            "gender": "女",
            "name": "赵畅",
            "_id": "CP4oZoh5JmHnKWs4Z",
            "organization": {
                "name": "大庆苹果集团",
                "_id": "vxu7QaYhLLQcwFDCZ"
            }
        },
        "members": [
            {
                "gender": "女",
                "name": "赵畅",
                "_id": "CP4oZoh5JmHnKWs4Z"
            },
            {
                "gender": "女",
                "name": "刘菲",
                "_id": "3ZRP9KzFsS9wHphqN"
            },
            {
                "gender": "男",
                "name": "高菲",
                "_id": "7M2MthqpiMry95Xs7"
            },
            {
                "gender": "女",
                "name": "向宏",
                "_id": "7SMEkxMfsNJsw5Dxz"
            },
            {
                "gender": "男",
                "name": "刘宝强",
                "_id": "LgKBonRAKme47g54t"
            },
            {
                "gender": "男",
                "name": "向博",
                "_id": "XWLNBorY2XLMxdxpT"
            },
            {
                "gender": "男",
                "name": "向朋友",
                "_id": "cDgGQvkJpostf3iTp"
            },
            {
                "gender": "女",
                "name": "周大海",
                "_id": "k8sDjGHKHhAQe8T9s"
            }
        ]
    },
```

**Comment by theodorDiaconu** [comment]
indeed, glad you figured it out.

**Comment by crapthings** [comment]
okay i think this could consider as a bug

members collection has 3 docs A, B, C

when you make link like

Projects owner = members' A

Projects members include members' A

your query

```
{
 owner: {
  field1,
},
members: {
field1: 1,
 field2: 1,
}
}
```

this will break futher reference

**Comment by theodorDiaconu** [comment]
@crapthings I don't understand what you mean by this:
```
Projects owner = members' A
Projects members include members' A
```

You mean that the owner is a member. If yes, how does it break, can you show me a sample response ?

**Comment by crapthings** [comment]
yes i will make a reproduce

**Comment by crapthings** [comment]
i meant



```

groupA: {
   name: 'A',
}

memberA: {
  name: 'A',
  gender: 'female',
  groupId: 'groupA',
}

project: {
   ownerId: 'memberA',
   memberIds: ['memberA']
}
```

when you query

```
{
  owner: {
    name: 1,
    organization: {
      name: 1,
    }
  },
  members: {
    name: 1,
    gender: 1,
  }
}
```

this will failed to fetch organization field,

to prevent this, you have to tell owner fetch  gender too,

because  package will not merge  this prop request

**Comment by theodorDiaconu** [comment]
@crapthings just tried your scenario locally with users, groups. It works as expected I'm sure your missing something :)

**Comment by crapthings** [comment]
```
git clone https://github.com/crapthings/app-architecture-proof.git
git checkout reproduce
meteor npm i
meteor
```

open this url then hit the reset fixture button inside bluebar, and watch the console.log

> http://localhost:3000/features/projects

**Comment by crapthings** [comment]
when u hitting reset fixture fastly

u will see "group 1" filcking on that empty field,  and  the console.log   has correctly document looks like

```
{
    "name": 1,
    "ownerId": 1,
    "owner": {
        "name": 1,
        "group": {
            "name": 1
        }
    },
    "memberIds": 1,
    "members": {
        "name": 1,
        "gender": 1
    }
}

[
    {
        "memberIds": [
            "HD8qYJcyG8Qiucvya"
        ],
        "name": "project 001",
        "ownerId": "HD8qYJcyG8Qiucvya",
        "_id": "jMp7PnFkpTFoJiMpA",
        "owner": {
            "name": "member 1",
            "_id": "HD8qYJcyG8Qiucvya",
            "group": {
                "name": "group 1",
                "_id": "KBQ7dovEwyqnjEqJA"
            }
        },
        "members": [
            {
                "name": "member 1",
                "_id": "HD8qYJcyG8Qiucvya"
            }
        ]
    }
]
```

**Comment by crapthings** [comment]
copy me to /app.js  to  see correctly result

```

export default {
  requireLogin: false,

  domains: [
    {
      name: 'home',
      isHome: true,
    },

    {
      name: 'projects',
      collection: new Mongo.Collection('projects'),
      fixture: 1,
      fields: [
        {
          name: 'name',
          type: String,
          onGrid: true,
          fake() {
            return 'project 001'
          },
        },

        {
          name: 'ownerId',
          type: String,
          ref: ['one', 'members', 'owner'],
          onGrid: true,
          fake: true,
        },

        {
          as: 'owner.name',
          onGrid: true,
        },

        { // comment me will failed
          as: 'owner.gender',
          onGrid: true,
        }, // comment me will failed

        {
          as: 'owner.group.name',
          onGrid: true,
        },

        {
          name: 'memberIds',
          type: [String],
          ref: ['many', 'members', 'members'],
          onGrid: true,
          fake: true,
        },

        {
          as: 'members.name',
          onGrid(d) {
            return _(d.members).map('name').join(', ')
          },
        },

        {
          as: 'members.gender',
          onGrid(d) {
            return _(d.members).map('gender').join(', ')
          },
        },
      ],
    },

    {
      name: 'members',
      collection: new Mongo.Collection('members'),
      fixture: 1,
      fields: [
        {
          name: 'name',
          type: String,
          onGrid: true,
          fake() {
            return 'member 1'
          },
        },

        {
          name: 'gender',
          type: String,
          onGrid: true,
          fake() {
            return 'male'
          }
        },

        {
          name: 'groupId',
          type: String,
          ref: ['one', 'groups', 'group'],
          onGrid: true,
          fake: true,
        },
      ],
    },

    {
      name: 'groups',
      collection: new Mongo.Collection('groups'),
      fixture: 1,
      fields: [
        {
          name: 'name',
          type: String,
          onGrid: true,
          fake() {
            return 'group 1'
          },
        },
      ],
    },
  ]

}

```

**Comment by theodorDiaconu** [comment]
@crapthings I see the problem, it's so freaking weird. I'm guessing it's somehow related to publish-composite.

**Comment by crapthings** [comment]
yes  and i can't reproduce with another simple example

https://github.com/crapthings/grapher-reproduce-101/blob/master/client/main.js

https://github.com/crapthings/grapher-reproduce-101/blob/master/lib/collections.js

**Comment by theodorDiaconu** [comment]
@crapthings it seems I found the issue. And it's stupid, and it's related to mergebox. Basically you will have 2 collisions on the same collection the owner and the members.

Now what happens behind the scenes, is this:
Because your query is reactive, the fetching is done client-side, not server-side. It does it client-side from local collections.

2 Cursors will try to push data for users, the one from ownerId, the one from memberIds. So the cursor from ownerId will push added elements. When the cursor from memberIds pushes the initial "added" elements. IF it sees that you already have that element, and if some fields aren't matching it will override your entry. 

So you will see in the websocket frames:
```
a["{\"msg\":\"changed\",\"collection\":\"members\",\"id\":\"vAfCgNWamFuYnyK2R\",\"fields\":{\"gender\":\"male\"},\"cleared\":[\"groupId\"]}"]
```

Because when the 2nd one does "added", it will try to compare it with the session view, and I believe it thinks something like, does this has any new fields ? If yes, then I'm taking charge. 

The one to blame here is "MergeBox" from Meteor. 

The solutions here are the following:
- Trick mergebox and keep the same fields
- Some new plan of publish composition that is aware of all the other collections


If you have a publication that returns:
```
[
   Users.find({}, {field1: 1}),
   Users.find({}, {field2: 1});
]
```

Or 2 publications like that. The last subscription will take ownership.

Sorry :(

**Comment by theodorDiaconu** [comment]
I will think about this in depth, but it's Meteor Mergebox who does this strange behavior.

**Comment by crapthings** [comment]
http://stackoverflow.com/questions/11751790/in-meteor-how-can-i-publish-one-server-side-mongo-collection-under-different-nam?rq=1#comment15629962_11764189
http://stackoverflow.com/a/12663746/1177684

https://github.com/versolearning/find-from-publication

https://github.com/yeputons/meteor-smart-publish

https://github.com/peerlibrary/meteor-control-mergebox

**Comment by alvestjo** [comment]
Hi,
This is very strange behavior of merge-box. Thanks for investigating!

Wouldn't it be possible to create a new client-side collection for each query?
I found this article:
[https://medium.com/@MaxDubrovin/workaround-for-meteor-limitations-if-you-want-to-sub-for-more-nested-fields-of-already-received-docs-eb3fdbfe4e07#.29tdexotu](https://medium.com/@MaxDubrovin/workaround-for-meteor-limitations-if-you-want-to-sub-for-more-nested-fields-of-already-received-docs-eb3fdbfe4e07#.29tdexotu)
I did not read it very closely yet...
//jonas
